### PR TITLE
STORM-1214: add javadoc for Trident Streams and Operations

### PR DESCRIPTION
**Pull request by ptgoetz**
JIRA: https://issues.apache.org/jira/browse/STORM-1214

This is just a work in progress for STORM-1214, if this is merged, please don't resolve the JIRA.

I plan on further docs. But I didn't want this to get lost in the 1.x work.

### CODE REVIEW

File: storm-core/src/jvm/org/apache/storm/trident/Stream.java [hunk 1]
```diff
@@ -68,61 +85,160 @@ protected Stream(TridentTopology topology, String name, Node node) {
...
     public Stream identityPartition() {
         return partition(new IdentityGrouping());
     }
-    
+
+    /**
+     * ## Repartitioning Operation
+     *
+     * This method takes in a custom partitioning function that implements
+     * {@link backtype.storm.grouping.CustomStreamGrouping}
```
Comment (satishd) [inline storm-core/src/jvm/org/apache/storm/trident/Stream.java]: It should be org.apache.storm.grouping.CustomStreamGrouping

File: storm-core/src/jvm/org/apache/storm/trident/operation/BaseOperation.java [hunk 2]
```diff
@@ -19,12 +19,27 @@
 
 import java.util.Map;
 
+/**
+ * Convenience implementation of the {@link storm.trident.operation.Operation} interface.
```
Comment (satishd) [inline storm-core/src/jvm/org/apache/storm/trident/operation/BaseOperation.java]: It should be org.apache.storm.trident.operation.Operation

File: storm-core/src/jvm/org/apache/storm/trident/operation/Operation.java [hunk 3]
```diff
@@ -20,7 +20,26 @@
...
+ * being shut down.
+ */
 public interface Operation extends Serializable {
+    /**
+     * Called when the `Operation` is first initialized.
+     * @param conf the Storm configuration map
+     * @param context the operation context which provides information such as the number of partitions in the stream,
+     *                and the current partition index. It also provides methods for registering operation-specific
+     *                metrics.
+     * @see storm.trident.operation.TridentOperationContext
```
Comment (satishd) [inline storm-core/src/jvm/org/apache/storm/trident/operation/Operation.java]: It should be org.apache.storm.trident.operation.TridentOperationContext

File: storm-core/src/jvm/org/apache/storm/trident/operation/Assembly.java [hunk 4]
```diff
@@ -20,6 +20,26 @@
...
+ * The `Assembly` interface provides a means to encapsulate logic applied to a {@link storm.trident.Stream}.
+ *
+ * Usage:
+ *
+ * ```java
+ * Stream mystream = ...;
+ * Stream assemblyStream = mystream.applyAssembly(myAssembly);
+ * ```
+ *
+ * @see storm.trident.Stream
```
Comment (satishd) [inline storm-core/src/jvm/org/apache/storm/trident/operation/Assembly.java]: It should be org.apache.storm.trident.Stream

File: storm-core/src/jvm/org/apache/storm/trident/operation/Assembly.java [hunk 5]
```diff
@@ -20,6 +20,26 @@
...
+ * @see storm.trident.operation.builtin.FirstN
```
Comment (satishd) [inline storm-core/src/jvm/org/apache/storm/trident/operation/Assembly.java]: org.apache.storm.trident.operation.builtin.FirstN

**Comment by satishd** [comment]
backtype.storm and storm.* should be replaced with org.apache.storm.* in all the updated javadoc.

File: storm-core/src/jvm/org/apache/storm/trident/Stream.java [hunk 6]
```diff
@@ -57,6 +57,23 @@
 import org.apache.storm.trident.state.StateUpdater;
 import org.apache.storm.trident.util.TridentUtils;
 
+/**
+ * A Stream represents the core data model in Trident, and can be thought of as a "stream" of tuples that are processed
+ * as a series of small batches. A stream is partitioned accross the nodes in the in the cluster, and operations are
```
Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/Stream.java]: `the nodes in the *in the* cluster` the second `in the` should probably go away.

File: storm-core/src/jvm/org/apache/storm/trident/Stream.java [hunk 7]
```diff
@@ -57,6 +57,23 @@
...
+ * applied to a stream in parallel accross each partition.
+ *
+ * There are five types of operations that can be performed on streams in Trident
+ *
+ * 1. **Partiton-Local Operations** - Operations that are applied locally to each partition and do not involve network
+ * transfer
+ * 2. **Repartitioning Operations** - Operations that that change how tuples are partitioned across tasks(thus causing
```
Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/Stream.java]: `Operations that *that* change` extra that

File: storm-core/src/jvm/org/apache/storm/trident/operation/Assembly.java [hunk 8]
```diff
@@ -20,6 +20,26 @@
 import org.apache.storm.trident.Stream;
 
 
+/**
+ * The `Assembly` interface provides a means to encapsulate logic applied to a {@link storm.trident.Stream}.
```
Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/operation/Assembly.java]: This one too should be org.apache....

File: storm-core/src/jvm/org/apache/storm/trident/operation/Filter.java [hunk 9]
```diff
@@ -19,6 +19,30 @@
...
+ *
+ * ### Configuration
+ * If your `Filter` implementation has configuration requirements, you will typically want to extend
+ * {@link storm.trident.operation.BaseFilter} and override the
+ * {@link storm.trident.operation.Operation#prepare(Map, TridentOperationContext)} method to perform your custom
+ * initialization.
+
+ *
+ * @see storm.trident.Stream
+ */
```
Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/operation/Filter.java]: all the links here need to be org.apache....

File: storm-core/src/jvm/org/apache/storm/trident/operation/Function.java [hunk 10]
```diff
@@ -19,6 +19,73 @@
...
+ * initialization.
+ *
+ * ### Performance Considerations
+ * Because Trident Functions perform logic on individual tuples -- as opposed to batches -- it is advisable
+ * to avoid expensive operations such as database operations in a Function, if possible. For data store interactions
+ * it is better to use a {@link storm.trident.state.State} or {@link storm.trident.state.QueryFunction} implementation
+ * since Trident states operate on batch partitions and can perform bulk updates to a database.
+ *
+ *
+ */
```
Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/operation/Function.java]: org.apache in the links here too.

File: storm-core/src/jvm/org/apache/storm/trident/operation/Operation.java [hunk 11]
```diff
@@ -20,7 +20,26 @@
...
+     */
     void prepare(Map conf, TridentOperationContext context);
+
+    /**
+     * When running in local mode, called when the local cluster is being shut down.
```
Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/operation/Operation.java]: Is this true for only local mode?  We made some changes so it should work in distributed mode, but it is best effort, and not guaranteed to happen.

File: storm-core/src/jvm/org/apache/storm/trident/operation/TridentCollector.java [hunk 12]
```diff
@@ -20,7 +20,33 @@
...
+ * Trident components that have the ability to emit tuples to a stream are passed an instance of this
+ * interface.
+ *
+ * For example, to emit a new tuple to a stream, you would do something like the following:
+ *
+ * ```java
+ *      collector.emit(new Values("a", "b", "c"));
+ * ```
+ * @see storm.trident.Stream
+ * @see backtype.storm.tuple.Values
```
Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/operation/TridentCollector.java]: Links need org.apache here too.

File: storm-core/src/jvm/org/apache/storm/trident/operation/TridentCollector.java [hunk 13]
```diff
@@ -20,7 +20,33 @@
...
+     * Emits a tuple to a Stream
+     * @param values a list of values of which the tuple will be composed
+     */
     void emit(List<Object> values);
+
+    /**
+     * Reports an error. The corresponding stack trace will be visible in the Storm UI.
+     *
+     * Note that calling this method does not alter the processing of a batch. To explicitly fail a batch and trigger
+     * a replay, components should throw {@link backtype.storm.topology.FailedException}.
```
Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/operation/TridentCollector.java]: Link here too

File: storm-core/src/jvm/org/apache/storm/trident/operation/builtin/FirstN.java [hunk 14]
```diff
@@ -28,6 +28,11 @@
 import org.apache.storm.trident.tuple.TridentTuple;
 
 
+/**
+ *
+ * An {@link storm.trident.operation.Assembly} implementation
```
Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/operation/builtin/FirstN.java]: org.apache.

**Comment by revans2** [comment]
Looks good over all.  Just a few comments about org.apache.storm instead or backtype.storm and storm.trident

Also one question about cleanup that I think needs to be updated.

**Comment by ptgoetz** [comment]
Sorry for the package renaming noise. I missed a commit.

File: storm-core/src/jvm/org/apache/storm/trident/operation/Operation.java (hunk 11 above)
Comment (ptgoetz) [inline storm-core/src/jvm/org/apache/storm/trident/operation/Operation.java]: Good catch. I'll fix this.

Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/operation/Operation.java]: ping? I am +1 after this is fixed

Comment (revans2) [inline storm-core/src/jvm/org/apache/storm/trident/operation/Operation.java]: Never mind this is fine.  I missed the fix somehow.

**Comment by revans2** [comment]
+1

**Comment by revans2** [comment]
@ptgoetz can I assume that you want this in 1.x as well?

**Comment by ptgoetz** [comment]
@revans2 Yes, please.

**Comment by revans2** [comment]
@ptgoetz I merged this to master and 1.x-branch but I am not going to resolve the JIRA because like you said you are not done yet, and there is more to do for this.

**Pull request merged**